feat: eme error interface

adrums86 opened the pull request:
This adds new error constants from video.js for a more cohesive errors interface, this also adds some new events for granular insight into the EME workflow.

dzianis-dashkevich commented on src/plugin.js:
```diff
@@ -188,9 +188,12 @@ export const setupSessions = (player) => {
 export const emeErrorHandler = (player) => {
   return (objOrErr) => {
+    const vjsErrorType = videojs.Error && videojs.Error.EMEKeySessionCreationError;
+    const errorType = vjsErrorType || 'eme-key-session-creation-error';
```
I guess we should not fallback.
We should expect `videojs.Error.EMEKeySessionCreationError` is available and we should update minimum videojs peerDependency version

dzianis-dashkevich commented:
Do we plan to include other errors?

adrums86 commented:
I added an error constant for each keysystem event and added videojs as a peerDependency. Changing this PR to a draft until the constants merge into video.js

adrums86 commented:
Can't flip this back to a draft PR for whatever reason. This shouldn't be merged until the corresponding video.js chore is merged and released first: https://github.com/videojs/video.js/pull/8634

dzianis-dashkevich commented on src/eme.js:
```diff
@@ -82,10 +82,14 @@ export const getSupportedKeySystem = (keySystems) => {
+      promise = promise.catch(() => {
+        // TODO: Not sure if we want to call emeError here or not.
+        window.navigator.requestMediaKeySystemAccess(keySystem, supportedConfigurations);
+      });
```
Here is just an accumulation of the promise,
I guess you should trigger error once in the other place

![CleanShot 2024-03-12 at 13 12 11@2x](https://github.com/videojs/videojs-contrib-eme/assets/98566601/96c8596c-c23b-4479-9d8e-6f3367bbce03)

dzianis-dashkevich commented on src/plugin.js:
```diff
@@ -241,8 +250,10 @@ const onPlayerReady = (player, emeError) => {
-        .catch(emeError);
+      handleEncryptedEvent(player, event, getOptions(player), player.eme.sessions, player.tech_, emeError)
+        .catch((error) => {
+          emeError(error, videojs.Error.EMEEncryptedError);
```
what is `videojs.Error.EMEEncryptedError`?

dzianis-dashkevich replied:
discussed offline

dzianis-dashkevich commented on src/plugin.js:
```diff
@@ -258,13 +269,17 @@ const onPlayerReady = (player, emeError) => {
     });
-    player.tech_.on('mskeyerror', emeError);
+    const msKeyErrorCallback = (error) => {
+      emeError(error, videojs.Error.MSKeyError);
```
Why do we have key system-specific errors?
I think we should use common interface for all of them

dzianis-dashkevich replied:
discussed offline

dzianis-dashkevich commented on src/plugin.js:
```diff
@@ -362,7 +377,9 @@ const eme = function(options = {}) {
-          .catch(emeError);
+          .catch((error) => {
+            emeError(error, videojs.Error.WebkitKeyError);
+          });
```
as we discussed offline:
we can analyze each rejection here, and try to use spec errors for legacy flow

dzianis-dashkevich commented on src/plugin.js:
```diff
@@ -334,7 +349,7 @@ const eme = function(options = {}) {
             callback(event.target.error);
             if (!suppressErrorIfPossible) {
-              emeError(event.message);
+              emeError(event.message, videojs.Error.MSKeyError);
```
as we discussed offline:
we can analyze each rejection here, and try to use spec errors for legacy flow

dzianis-dashkevich commented on src/plugin.js:
```diff
@@ -319,12 +334,12 @@ const eme = function(options = {}) {
             callback(error);
             if (!suppressErrorIfPossible) {
-              emeError(error);
+              emeError(error, videojs.Error.EMEEncryptedError);
```
as discussed offline:
we should probably analyze possible rejections here to narrow it to specific error or remove it
